Refactor sudt with registry address

jjyr opened the pull request:
[CHANGES](https://github.com/jjyr/godwoken/blob/refactor-sudt-with-registry-address/docs/CHANGES.md)

## Related PRs
- https://github.com/nervosnetwork/godwoken-scripts/pull/116
- https://github.com/nervosnetwork/godwoken-polyjuice/pull/135

jjyr committed:
refactor: Support registry address

jjyr committed:
refactor: Add registry_id to builtin scripts structure

jjyr committed:
refactor: fix gw-common reference

jjyr committed:
refactor: Add registry_id to DepositLockArgs

jjyr committed:
feat: Handling EOA ETH address mapping in deposit

jjyr committed:
feat: try to extract eth address before handle deposit

jjyr committed:
fix: fix address len check

jjyr committed:
fix: fix RegistryAddress#write_to_slice

jjyr committed:
fix: SYS_PAY_FEE

jjyr committed:
fix: accept parse registry address from a longer slice

jjyr committed:
feat: Add eth registry as a builtin script

jjyr committed:
fix: rebase gw-tools

jjyr committed:
fix: fix toml value after table error

jjyr committed:
feat: Using registry address RPC to replace short script hash RPC

jjyr committed:
doc: docs/CHANGES.md

jjyr committed:
refactor: change withdrawal EIP712 signature structure

jjyr committed:
feat: using EIP712 to sign L2Transaction

jjyr committed:
feat: update CCWithdrawalWitness

jjyr committed:
chore: derive Debug on EIP-712 structs

jjyr committed:
feat: Using RollupConfig#chain_id as Ethereum tx chain id

jjyr committed:
chore: update tests binaries

jjyr committed:
docs: update CHANGES.md

blckngm commented on docs/CHANGES.md:
```diff
@@ -0,0 +1,228 @@
+# v1.0 CHANGES
+
+In Godwoken v0, Godwoken EVM has using an internal address format, users need to integrate web3-provider to handling the conversion between Godwoken internal address and the Ethereum address. Even in most cases the conversion is handled transparently, this incompatiblity still causes some problem in conner cases. Such as, we cannot hanlde the conversion happend inside metamask or some hardware wa ...
```
```suggestion
In Godwoken v0, Godwoken EVM was using an internal address format. Users need to integrate web3-provider to handling the conversion between the internal address format and the Ethereum address format. Although in most cases the conversion is handled transparently, this incompatibility still causes problems in some corner cases. Such as, we cannot handle the conversion in metamask or some hardware wallets because the web3-provider plugin is unsupported in these environments.
```

blckngm commented on docs/CHANGES.md:
```diff
@@ -0,0 +1,228 @@
+
+In Godwoken v0, Godwoken EVM has using an internal address format, users need to integrate web3-provider to handling the conversion between Godwoken internal address and the Ethereum address. Even in most cases the conversion is handled transparently, this incompatiblity still causes some problem in conner cases. Such as, we cannot hanlde the conversion happend inside metamask or some hardware wa ...
+
+In the v1 version, we add a new builtin contract: Ethereum address registry, when a new account is created by user deposit, an Ethereum address is inserted to the contract. With the new builtin contract, the Ethereum address is supported in the EVM directly, now users can submit their transaction to Godwoken without using the web3-provider plugin. We suppose to provide 100% compatibility from the ...
```
```suggestion
In the v1 version, we add a new builtin contract: Ethereum address registry. When a new account is created by user deposit, an Ethereum address is inserted to the contract. With the new builtin contract, the Ethereum address is supported in the EVM directly. Now users can submit their transaction to Godwoken without using the web3-provider plugin. We aim to provide 100% compatibility at the RPC level. When users switch to the Godwoken network, they can use Ethereum Dapps or toolchains without modification.
```

Also, would it be too strong to claim “without modification”? Maybe “with minimal modifications”?

blckngm commented on docs/CHANGES.md:
```diff
@@ -0,0 +1,228 @@
+
+### Core data structure
+
+* `RollupConfig#compatible_chain_id` is changed to `RollupConfig#chain_id`. In v1 we directly using `RollupConfig#chain_id` as the `chain_id` in the signature.
```
```suggestion
* `RollupConfig#compatible_chain_id` is changed to `RollupConfig#chain_id`. In v1 we use `RollupConfig#chain_id` as the `chain_id` in the signature directly.
```

blckngm commented on docs/CHANGES.md:
````diff
@@ -0,0 +1,228 @@
+}
+```
+
+* A field `registry_id` is added on the `DepositLockArgs` & `DepositRequest` to indicate which registry we deposited to (The `registry_id` should always using the builtin Ethereum address registry in the v1).
````
```suggestion
* A field `registry_id` is added on the `DepositLockArgs` & `DepositRequest` to indicate which registry we deposited to (The `registry_id` should always be the builtin Ethereum address registry in v1).
```

blckngm commented on docs/CHANGES.md:
````diff
@@ -0,0 +1,228 @@
+}
+```
+
+* The `RawWithdrawalRequest`, a field `registry_id` is added on the structure to indicate which registry it withdraws from.
````
```suggestion
* In `RawWithdrawalRequest`, a field `registry_id` is added to indicate which registry it withdraws from.
```

blckngm commented on docs/CHANGES.md:
````diff
@@ -0,0 +1,228 @@
+}
+```
+
+* Pay fee with Simple UDT is removed in the v1, `registry_id` is added to `Fee` to indicate which registry the token in transfered to.
````
```suggestion
* Paying fee with Simple UDT is removed in v1, `registry_id` is added to `Fee` to indicate which registry the token is transfered to.
```

blckngm commented on docs/CHANGES.md:
```diff
@@ -0,0 +1,228 @@
+
+### Address registry
+
+1. A new builtin contract Ethereum address registry is added, Godwoken using it to handling the Ethereum address. When user deposit token to create a new account, a corresponded Ethereum address of the new account is inserted to the contract.
```
```suggestion
1. A new builtin contract Ethereum address registry is added. Godwoken uses it to handle the Ethereum address format. When a user deposits token to create a new account, a corresponding Ethereum address is inserted to the contract.
```

blckngm commented on docs/CHANGES.md:
```diff
@@ -0,0 +1,228 @@
+### Address registry
+
+1. A new builtin contract Ethereum address registry is added, Godwoken using it to handling the Ethereum address. When user deposit token to create a new account, a corresponded Ethereum address of the new account is inserted to the contract.
+2. Deposit automatically mapping the Ethereum address for new accounts. If the account is created through Meta contract, the developer must register the Ethereum address for the acount by calling the Ethereum address registry contract.
```
```suggestion
2. Deposit automatically maps the Ethereum addresses for new accounts. If the account is created through a Meta contract, the contract developer must register the Ethereum address for the acount by calling the Ethereum address registry contract.
```

blckngm commented on docs/CHANGES.md:
```diff
@@ -0,0 +1,228 @@
+
+### Signing
+
+In v0 we only support sign transactions with personal sign, this is considered insecure since users can only see a random 32-bytes hex, they have no idea what data they are signing.
```
```suggestion
In v0 we only support signing transactions with personal sign, this is considered insecure since users can only see a random 32-bytes hex, they have no idea what data they are signing.
```

blckngm commented on docs/CHANGES.md:
```diff
@@ -0,0 +1,228 @@
+
+In v0 we only support sign transactions with personal sign, this is considered insecure since users can only see a random 32-bytes hex, they have no idea what data they are signing.
+
+V1 support two signature format:
```
```suggestion
V1 supports two signature formats:
```

blckngm commented on docs/CHANGES.md:
```diff
@@ -0,0 +1,228 @@
+
+V1 support two signature format:
+
+1. The Ethereum transaction format, users can signing transaction in metamask with the same experience in the ethereum world.
```
```suggestion
1. The Ethereum transaction format, users can sign transactions in Metamask with the same experience in the ethereum world.
```

blckngm commented on docs/CHANGES.md:
```diff
@@ -0,0 +1,228 @@
+V1 support two signature format:
+
+1. The Ethereum transaction format, users can signing transaction in metamask with the same experience in the ethereum world.
+2. The EIP-712 format, we support using EIP-712 format to sign withdrawal message, or a Godwoken transaction, users can check the content of the transaction before they are signing.
```
```suggestion
2. The EIP-712 format. We support using EIP-712 format to sign withdrawal messages or Godwoken transactions. Users can check the content of a transaction before they sign.
```

Flouse committed:
fix(tools): get_balance_by_script_hash failed at thhe first time

Flouse committed:
Bump Godwoken version to v1.1.0

Flouse committed:
Deprecate EOA mapping crate

Flouse committed:
chore: update the refs of integration-test

Flouse committed:
Refactor create_creator_account in gw-tools

Flouse committed:
fix(tool): use EIP-712 typed message in `create_creator_account`

ErrMsg:
1. failed Unlock error Invalid signature: Wrong chain_id, expected: 71400 actual: 0
2. Invalid signature: Mismatch pubkey hash

jjyr committed:
docs: update CHANGES.md

Co-authored-by: Yin Guanhao <sopium@users.noreply.github.com>

jjyr committed:
chore: remove tron account lock

blckngm commented on crates/common/src/registry_address.rs:
```diff
@@ -0,0 +1,58 @@
+        }
+        let registry_id = u32::from_le_bytes(slice[..4].try_into().unwrap());
+        let address_len = u32::from_le_bytes(slice[4..8].try_into().unwrap());
+        if slice.len() < (8 + address_len) as usize {
```
`8 + address_len` can panic when overflowing.

Better written as this:

```
let address_len: usize = u32::from_le_bytes(slice[4..8].try_into().unwrap()).try_into().ok()?;
if slice.len() - 8 < address_len {
    ...
    address: slice[8..][..address_len].to_vec(),
```

blckngm replied:
Or with `bytes::Buf`:

```rust
    pub fn from_slice(mut slice: &[u8]) -> Option<Self> {
        if slice.len() < 8 {
            return None;
        }
        let registry_id = slice.get_u32_le();
        let address_len: usize = slice.get_u32_le().try_into().ok()?;
        if slice.len() < address_len {
            return None;
        }
        let reg_addr = RegistryAddress {
            registry_id,
            address: slice[..address_len].to_vec(),
        };
        Some(reg_addr)
    }
```

blckngm commented on crates/common/src/registry_address.rs:
```diff
@@ -0,0 +1,58 @@
+            return Err(self.len());
+        }
+        buf[..4].copy_from_slice(&self.registry_id.to_le_bytes());
+        buf[4..8].copy_from_slice(&(self.address.len() as u32).to_le_bytes());
```
`as u32` may truncate the length value.

blckngm replied:
Actually we should probably set an address length limit much smaller than `u32::MAX_VALUE`.

jjyr committed:
fix: check registry address len to prevent panic

zeroqn commented on crates/generator/src/account_lock_manage/eip712/types.rs:
```diff
@@ -64,9 +71,128 @@ impl EIP712Encode for WithdrawalAsset {
+    }
+
+    fn encode_type(&self, buf: &mut Vec<u8>) {
+        buf.extend(b"RegistryAddress(string registry,address address)");
```
string registry? uint?

jjyr replied:
We use string in signature, so users can understand the signing content.

jjyr committed:
fix: fix jsonrpc types

jjyr committed:
chore: add arg block-producer-address to generate-config command

jjyr committed:
chore: update kicker ref in integration tests

jjyr committed:
chore: rebase develop

jjyr commented:
Rebased

keroro520 commented on crates/types/src/offchain/pool.rs:
```diff
@@ -6,7 +7,7 @@ use super::DepositInfo;
 pub struct BlockParam {
     pub number: u64,
-    pub block_producer_id: u32,
+    pub block_producer: Bytes,
```
Is it possible to use `block_producer: RegistryAddress`? `Bytes` seems not friendly for readability.

jjyr replied:
You are right, but this value is used to fill the `L2Block#block_producer` field, so if we use `RegistryAddress` we need an extra serialization. IMO we can accept to use `Bytes` here since this structure is only used to generate a block.

keroro520 commented on crates/types/schemas/godwoken.mol:
```diff
@@ -238,16 +253,16 @@ union SUDTArgs {
-    // godwoken short script hash (script_hash[0..n])
-    to: Bytes,
+    // Godwoken registry address: (registry_id (4 bytes) | address len(4 bytes) | address)
+    to_address: Bytes,
```
I think `to` is enough. `to` always be an address.

jjyr replied:
Since we have `to_id` in the `L2Tx` structure, the `to_address` is used to emphasis the difference, to void the confuse between `to` & `to_id`.

jjyr committed:
fix: fix bench tests

keroro520 reviewed (dismissed):
LGTM

blckngm commented:
FeeQueue benchmark test is failing. Otherwise LGTM.

jjyr committed:
fix: fix fee_queue benchmark

jjyr commented:
Fixed the benchmark test. @sopium @keroro520

jjyr committed:
chore: upgrade integration tests kicker_ref to v1.1

zeroqn commented on crates/tools/src/account.rs:
```diff
@@ -81,55 +80,14 @@ pub fn privkey_to_l2_script_hash(
-        let r = GwBytes::from(hex::decode(account[2..].as_bytes())?);
+) -> Result<H256> {
+    // if match script hash
+    if account.starts_with("0x") && account.len() == 64 {
```
66?

jjyr replied:
Good catch

jjyr committed:
fix: fix parse_account_from_str
